### PR TITLE
Floor and Ceil

### PR DESCRIPTION
**Pull request by arn4**
## Proposed changes

In the code I've found some TODO referring to the fact that `floor` and `ceil` operations are needed. I've implemented that, including Metal support and Accelerate fallback, as first thing to get the hands on the project. It would be nice to a review from one of the maintainers. 

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -13,7 +13,7 @@ set(CMAKE_INSTALL_MESSAGE NEVER)
 option(MLX_BUILD_TESTS "Build tests for mlx" ON)
 option(MLX_BUILD_EXAMPLES "Build examples for mlx" ON)
 option(MLX_BUILD_BENCHMARKS "Build benchmarks for mlx" OFF)
-option(MLX_BUILD_PYTHON_BINDINGS "Build python bindings for mlx" OFF)
```
Comment (awni) [inline CMakeLists.txt:16]: Could you revert this? Not sure what the intention is.. but we don't build them by default intentionally

File: mlx/backend/common/primitives.cpp [hunk 2]
```diff
@@ -167,6 +167,17 @@ void Broadcast::eval(const std::vector<array>& inputs, array& out) {
   out.copy_shared_buffer(in, strides, flags, in.data_size());
 }
 
+void Ceil::eval(const std::vector<array>& inputs, array& out) {
+  assert(inputs.size() == 1);
+  auto& in = inputs[0];
+  if (not is_integral(in.dtype())) {
+    unary_fp(in, out, [](auto x) { return std::ceil(x); });
+  } else {
+    // No-op non-integer types
```
Comment (awni) [inline mlx/backend/common/primitives.cpp:176]: Nit: comment should say `No-op on integer types`

File: mlx/backend/common/primitives.cpp [hunk 3]
```diff
@@ -287,6 +298,17 @@ void Exp::eval(const std::vector<array>& inputs, array& out) {
   }
 }
 
+void Floor::eval(const std::vector<array>& inputs, array& out) {
+  assert(inputs.size() == 1);
+  auto& in = inputs[0];
+  if (not is_integral(in.dtype())) {
+    unary_fp(in, out, [](auto x) { return std::floor(x); });
+  } else {
+    // No-op non-integer types
```
Comment (awni) [inline mlx/backend/common/primitives.cpp:307]: Same

File: mlx/ops.cpp [hunk 4]
```diff
@@ -1498,6 +1498,15 @@ array minimum(const array& a, const array& b, StreamOrDevice s /* = {} */) {
...
 }
 
+array floor(const array& a, StreamOrDevice s /* = {} */) {
+  return array(
+      a.shape(), a.dtype(), std::make_unique<Floor>(to_stream(s)), {a});
+}
+
+array ceil(const array& a, StreamOrDevice s /* = {} */) {
+  return array(a.shape(), a.dtype(), std::make_unique<Ceil>(to_stream(s)), {a});
+}
```
Comment (awni) [inline mlx/ops.cpp:1508]: Could you throw an `std::invalid_argument` if the dtype is complex? (Same behavior as numpy)

**Review by awni — CHANGES_REQUESTED** [review]
This is beautifully done, thank you! Just a few minor comments and we can merge it!

**Review by angeloskath — CHANGES_REQUESTED** [review]
This is fantastic! Thanks. I left a comment and together with Awni's I think this is a great addition.

File: CMakeLists.txt (hunk 1 above)
Comment (arn4) [reply CMakeLists.txt:16]: I struggled a bit with cmake at first, so I was playing the settings to understand what to do. I forgot to revert before committing

**Comment by arn4** [comment]
Thanks for the comments! I've implemented all your reviews.

**Comment by angeloskath** [comment]
Looks great.

Could you also fix the upcast in the metal kernel? ie remove the cast to float and then back to the original dtype.

**Comment by arn4** [comment]
Done!

**Review by awni — APPROVED** [review]
Thanks this is awesome 🚀

**Pull request merged**